fabric-brewing-v1/src/main/java/net/fabricmc/fabric/impl/brewing/BrewingRecipe.java

```diff
@@ -0,0 +1,106 @@
+import net.minecraft.client.MinecraftClient;
+import net.minecraft.inventory.Inventory;
+import net.minecraft.item.ItemStack;
+import net.minecraft.recipe.Ingredient;
+import net.minecraft.recipe.Recipe;
+import net.minecraft.recipe.RecipeSerializer;
+import net.minecraft.recipe.RecipeType;
+import net.minecraft.util.DefaultedList;
+import net.minecraft.util.Identifier;
+import net.minecraft.world.World;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class BrewingRecipe implements Recipe<Inventory> {
+    private final Identifier id;
+    private final String group;
+    private final PotionIngredient input;
+    private final PotionIngredient basePotion;
+    private final PotionIngredient output;
+    private List<Integer> matchedSlots = new ArrayList<>();
+
+    public BrewingRecipe(Identifier id, String group, PotionIngredient input, PotionIngredient basePotion, PotionIngredient output) {
+        this.id = id;
+        this.group = group;
+        this.input = input;
+        this.basePotion = basePotion;
+        this.output = output;
+    }
+
+    public boolean matches(Inventory inv, World world) {
+        boolean matchesPotion = false;
+        if(basePotion.test(inv.getInvStack(0))) { matchesPotion = true; this.matchedSlots.add(0); };
+        if(basePotion.test(inv.getInvStack(1))) { matchesPotion = true; this.matchedSlots.add(1); };
+        if(basePotion.test(inv.getInvStack(2))) { matchesPotion = true; this.matchedSlots.add(2); };
+        return input.test(inv.getInvStack(3)) && matchesPotion;
+    }
+
+    public ItemStack craft(Inventory inv) { return output.asIngredient().getStackArray()[0].copy(); }
+    public boolean fits(int w, int h) { return true; }
+    public PotionIngredient getInput() { return input; }
+    public PotionIngredient getBasePotion() { return basePotion; }
+    public ItemStack getOutput() { return output.asIngredient().getStackArray()[0]; }
+    public RecipeType<?> getType() { return FabricBrewingInit.BREWING_RECIPE_TYPE; }
+    public Identifier getId() { return id; }
+    public RecipeSerializer<?> getSerializer() { return FabricBrewingInit.BREWING_RECIPE_SERIALIZER; }
+    public String getGroup() { return group; }
+
+    public PotionIngredient getOutputPotionIngredient() { return output; }
+    public PotionIngredient.Type getOutputPotionType() { return output.getPotionType(); }
+    public List<Integer> getMatchedSlots() { return matchedSlots; }
+
+    public DefaultedList<Ingredient> getPreviewInputs() {
+        DefaultedList<Ingredient> list = DefaultedList.create();
+        list.add(input.asIngredient());
+        list.add(basePotion.asIngredient());
+        return list;
+    }
+
+    public static List<BrewingRecipe> getRelevantRecipes(Inventory inv, ItemStack stack, int slot) {
+        World world = inv instanceof BrewingStandBlockEntity
```

Comment:
Consider modded inventories.

Reply:
What would you suggest as a better way to manage this? `instanceof BlockEntity`? 

Also, this method is primarily a utility for working around the existing vanilla code, and shouldn't generally need to be called from a custom block entity that doesn't have to work around that.

Reply:
I suggest another method that allows mods to pass in a custom world instance instead of guessing wildly.

Reply:
Alright, I'll add a version that includes a `world` parameter as well.